gridded_etl_tools/utils/extractor.py

```diff
@@ -0,0 +1,373 @@
+
+import pathlib
+import typing
+import ftplib
+import re
+import time
+import multiprocessing
+
+
+class Extractor:
+    """
+    Base class for common extraction functions. Associate this class with a `DatasetManager` object by passing the
+    object at initialization.
+
+    Extractor's functions are used to request source data, using multiprocessing when possible. `Extractor.pool` is
+    used to launch the requests using a thread pool.
+
+    If more specific types of extraction are available in a child class, consider using the child class instead. For
+    example, if extracting from FTP, use FTPExtractor.
+    """
+
+    def __init__(self, dm: dataset_manager.DatasetManager):
+        """
+        Create a new Extractor object by associating a Dataset Manager with it.
+
+        Parameters
+        ----------
+        dm
+            Source data for this dataset manager will be extracted
+        """
+        self.dm = dm
+
+    def pool(self, batch_processor: typing.Callable[..., bool], batch: typing.Sequence[typing.Sequence]) -> bool:
+        """
+        Launch a batch of requests simultaneously, wait for them all to complete, then return a boolean indicating
+        whether any of the requests were successful.
+
+        The batch processor is any generic callable that accepts the arguments provided in each entry of the batch.
+        Each entry in the batch is a list of arguments to be forwarded to the batch processor. The batch is submitted
+        to a thread pool which will call the batch processor on every entry of the batch, launching all the calls in
+        parallel and waiting for them to complete.
+
+        If any of the calls returns `True`, this will return `True`, otherwise it will return `False`.
+
+        Parameters
+        ----------
+        batch_processor
+            Function which will be called repeatedly, being passed a new entry in the batch each time it's called
+        batch
+            A 2D list of lists (or any sequential type) containing arguments to be forwarded to the given batch
+            processor
+
+        Returns
+        -------
+        bool
+            `True` if any of the batch calls returns `True`, `False` otherwise
+        """
+        # Success remains false until the first successful request is completed
+        success = False
+
+        # Time the download
+        start_downloading = time.time()
+
+        # Thread count is one less than the amount of CPU available, unless there is only 1 CPU
+        thread_count = max(1, multiprocessing.cpu_count() - 1)
+
+        # run downloads
+        if len(batch) > 0:
+            self.dm.info(f"Submitting request for {len(batch)} datasets in parallel using {thread_count} threads.")
+
+            # download in parallel
+            with multiprocessing.pool.ThreadPool(processes=thread_count) as pool:
+                for result in pool.starmap(batch_processor, batch):
+                    if result:
```

Comment:
If any download succeeds the operation is considered as having succeeded? The operation only fails if all downloads fail?